File: static/js/notice/notice_settings.js
let app = new Vue({
    el: "#main",
    delimiters: ["[[", "]]"],
    data(){
        return {
            settings: [],
            _settings: [],
            ok: false
        }
    },
    created(){
        this.getSettings();
        // this.settings = [{'notice_type_id': 1, 'status': true, 'name': '客户保有量每日提醒', 'comment': '统计名下客户数量'}, 
        // {'notice_type_id': 2, 'status': true, 'name': '客户超限自动踢回公海池通知', 'comment': '超过限制自动踢回'}, 
        // {'notice_type_id': 3, 'status': false, 'name': '回款计划异常提醒', 'comment': '主要判断承诺回款日期填写是否异常'}, 
        // {'notice_type_id': 4, 'status': false, 'name': '账号开通提醒', 'comment': '快速开通秘书、指挥平台、网评、态势感知'}, 
        // {'notice_type_id': 5, 'status': 1, 'name': '秘书帐号重新关联商机每周一提醒', 'comment': '关联商机赢率达到100%，需要重新关联商机'}, 
        // {'notice_type_id': 6, 'status': 1, 'name': '任务提醒', 'comment': '商务的10种任务类型提醒'}, 
        // {'notice_type_id': 7, 'status': 1, 'name': '下载移动端和绑定微信公众号提醒', 'comment': '商务提醒客户下载移动端和绑定微信号'}, 
        // {'notice_type_id': 8, 'status': 1, 'name': '计划回款提醒', 'comment': '在计划回款日的前15天和30天提醒'}, 
        // {'notice_type_id': 9, 'status': 1, 'name': '订单到期提醒', 'comment': '订单到期前25天通知'}, 
        // {'notice_type_id': 10, 'status': 1, 'name': '停用账号活跃提醒', 'comment': '近30天内停用状态超过21天，但30天内活跃有6天'}];
        // this.ok = true;
    },

    methods: {
        getSettings(){
            let url = "/notice/setting/list/api";
            axios.get(url).then(res => {
                let settings = [];
                for (let setting of res.data.data){
                    setting.status = setting.status == 1 ? true : false;
                    settings.push(setting);
                };
                this.settings = settings;
                this.ok = true;
            }).catch(res => {
                console.log(res);
            })
        },

        changeSwitch(setting){
            let self = this;
            let url = "/notice/settings/change";
            let _setting = JSON.parse(JSON.stringify(setting));

            if (setting.status == 1){
                _setting.status = 1
            }else{
                _setting.status = 0
            }

            axios.get(url, {
                params: _setting
            }).then(res => {
                if (res.data.code !== 1){
                    self.$message.error(res.data.error);
                }
            }).catch(err => {
                self.$message.error(err);
            });
        },
    }
})

File: notice/views/notice_settings.py
import json
import logging
from django.http import JsonResponse
from django.shortcuts import render, redirect
from mandala.auth import get_user_model
from mandala.auth.decorators import permission_required,login_required
from notice.models import NoticeType, NoticeSetting
from notice.forms import  NoticeSettingForm
from user_center.models import UserLog

from django.conf import settings
URL_403 = settings.URL_403
JSON_403 = settings.JSON_403
# Create your views here.

logger = logging.getLogger("notice")
User = get_user_model()
@login_required()
@permission_required("notice.notice_setting.change",login_url=JSON_403)
def change_notice_setting(request):
    result = {"code": 0, "error": ""}
    if request.user.is_anonymous:
        result["error"] = "会话已过期，请重新登录"
        return JsonResponse(result)
    if request.method == "GET":
        kws = request.GET
    else:
        kws = request.POST

    ding_user_id = kws.get("ding_user_id")
    if not ding_user_id:
        ding_user_id = request.user.dtalkid
    if not request.user.is_superuser:
        if ding_user_id != request.user.dtalkid:
            result["error"] = "您没有权限"
            return JsonResponse(result)
    form = NoticeSettingForm(kws)
    if not form.is_valid():
        result["code"] = -1
        result["error"] = "参数错误"
        return JsonResponse(result)
    params = {"ding_user__dtalkid": ding_user_id, "notice_type_id": kws.get("notice_type_id")}
    ns = NoticeSetting.objects.filter(**params)
    if len(ns) > 0:
        ns.update(status=kws.get("status"))
    else:
        params = {"ding_user_id": ding_user_id, "notice_type_id": kws.get("notice_type_id"), "status": kws.get("status")}
        NoticeSetting.objects.create(**params)

    message = json.dumps(dict(kws))
    UserLog.objects.create(username=request.user.username, user_id=request.user.id,
            model="通知管理-通知设置", action="修改", message=message)

    result["code"] = 1
    result["error"] = "设置成功"
    return JsonResponse(result)

def get_notice_setting_list(ding_user_id):
    notice_type_list = list(NoticeType.objects.all().values())
    nsl = list(NoticeSetting.objects.filter(ding_user__dtalkid=ding_user_id).values())
    notice_setting_dict = {ns["notice_type_id"]: ns for ns in nsl}
    notice_setting_list = []
    # fields = NoticeSetting._meta.fields
    for notice_type in notice_type_list:
        notice_type_id = notice_type["id"]
        ns = notice_setting_dict.get(notice_type_id)
        if not ns:
            ns = {"notice_type_id": notice_type_id, "status": 1, "ding_user_id": ding_user_id}
        for k, v in notice_type.items():
            if k != "id":
                ns[k] = v
        notice_setting_list.append(ns)
    result = {"code": 1, "data": notice_setting_list, "error": ""}
    return result

@login_required()
@permission_required("notice.notice_setting.view",login_url=JSON_403)
def notice_setting_list_api(request):

    message = json.dumps(dict(request.GET))
    UserLog.objects.create(username=request.user.username, user_id=request.user.id,
            model="通知管理-通知设置api", action="查询", message=message)

    result = {"code": 0, "data": None, "error": ""}
    if request.method != "GET":
        result["code"] = -1
        result["error"] = "错误的方法"
        return JsonResponse(result)
    ding_user_id = request.GET.get("ding_user_id")
    if ding_user_id:
        if not request.user.is_superuser:
            return redirect(request.path)
    if not ding_user_id:
        ding_user_id = request.user.dtalkid
    result = get_notice_setting_list(ding_user_id)
    return JsonResponse(result)

@login_required()
@permission_required("notice.notice_setting.view",login_url=URL_403)
def notice_settings(request):

    message = json.dumps(dict(request.GET))
    UserLog.objects.create(username=request.user.username, user_id=request.user.id,
            model="通知管理-通知设置", action="访问", message=message)

    title = "修改钉钉通知设置"
    return render(request, "notice/notice_settings.html", locals())

                